### DISCUSSION TITLE
Q & A is it possible to add/symlink external binaries?

### DISCUSSION THREAD

**Discussion by nickchomey** [aquaproj/aqua#2953] [Deprecated (Question and Answer)]
### Question

I have some applications that build their binary via a Makefile. I'd like to use aqua to manage these binaries so that different projects can easily change between the various versions/branches

Is there any way I can add or symlink these binaries into aqua's local repository? If not, what options might be available to me?

**Comment by suzuki-shunsuke** [comment]
I think there is no way to support the use case.
aqua doesn't support executing external commands such as `make` except for some exceptions such as `go install`, `go build`, and `cargo install`.

https://aquaproj.github.io/docs/reference/restriction/#aqua-doesnt-support-running-any-external-commands-to-install-tools

> If not, what options might be available to me?

I have no idea.
aqua manages versions strictly, so changing versions by `branch` is unsupported.

**Reply by nickchomey** [reply]
Thanks for the quick reply!

I don't need aqua to execute external commands. I can run make on my own to build the binary. I just need a way of adding a binary that already exists in the filesystem to Aqua. 

Even if aqua doesn't have such a mechanism, surely there's a way to manually create a directory and symlink the binary, such that it would be recognized and used by aqua? (though an aqua command like `aqua link tool_name@version binary/path` would be fantastic)

It would be very useful for doing rapid local development and testing of tools, which could also be added to the oficial registry later. (this issue is in reference to ddev, which is in the aqua registry, but is built via makefile) 

In fact, in theory, this use case could apply to adding binaries that are installed through other means - eg Ubuntu apt package manager.

**Reply by suzuki-shunsuke** [reply]
I see. Thank you for your explanation.
If we support the use case, one option is to add a new [package type](https://aquaproj.github.io/docs/reference/registry-config/#package-types) `local_file`.

e.g.

registry.yaml

```yaml
packages:
- type: local_file
  name: foo/bar
  path: bin/bar-{{.Version}}
```

aqua.yaml

```yaml
packages:
- name: foo/bar@main
```

1. You install tools into specific paths somehow (aqua doesn't install them)
1. `aqua i` creates a symbolic link to `AQUA_ROOT_DIR/bin`
1. When the command is executed, aqua searches `bin/bar-{{.Version}}` and executes it. If the command isn't found, the command fails

I came up with this idea, but honestly I don't really want to.

**Reply by suzuki-shunsuke** [reply]
> In fact, in theory, this use case could apply to adding binaries that are installed through other means - eg Ubuntu apt package manager.

Is it possible to install multiple versions using apt?

**Reply by nickchomey** [reply]
I do think what you proposed there would be very useful, but there's no need to implement it if you don't want to!

But do you think there's a way that I can just manually add directories and a symlink to wherever aqua stores and tracks its tools and their versions?

As for installing multiple versions via apt, no I don't think it's possible. The thought was that we typically install plenty of binaries via apt, but sometimes want to customize the binary. With what I'm proposing (whether I do it manually or a `aqua link` command is created) there could be a tool@apt version, and tool@custom_branch, and others. All of which just symlink to wherever the binary is. 

But, to be clear, apt is not important here - the concept is just to add existing local binaries to Aqua, no matter how they got to the filesystem.
